src/main/java/com/khutircraftubackend/product/ProductController.java

```diff
@@ -1,23 +1,79 @@
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.multipart.MultipartFile;
 
+import java.io.IOException;
 import java.util.ArrayList;
 import java.util.List;
 
 @RestController
-@RequestMapping("/product")
+@RequestMapping("/products")
 @Slf4j
+@RequiredArgsConstructor
 public class ProductController {
+    private final ProductService productService;
 
+    @PostMapping("/create")
+    @PreAuthorize("hasRole('SELLER')")
+    public ResponseEntity<ProductEntity> createProduct(
+            @ModelAttribute ProductCreateRequest request,
+            @RequestParam("thumbnailImage")MultipartFile thumbnailImage,
+            @RequestParam("image") MultipartFile image) throws IOException {
+        ProductEntity newProduct = productService.createProduct(request, thumbnailImage, image);
 
+        return new ResponseEntity<>(newProduct, HttpStatus.CREATED);
+    }
+
+    @PatchMapping("/patch-product/{productId}")
+    @PreAuthorize("hasRole('SELLER') and @productService.canModifyProduct(#productId)")
+    public ResponseEntity<ProductEntity> patchProduct(@PathVariable Long productId,
+                                                      @ModelAttribute ProductUpdateRequest request,
+                                                      @RequestParam("thumbnailImage")MultipartFile thumbnailImage,
+                                                      @RequestParam("image") MultipartFile image) throws IOException {
+        ProductEntity patchedProduct = productService.patchProduct(productId, request, thumbnailImage, image);
+
+        return ResponseEntity.ok(patchedProduct);
+    }
+
+    @PutMapping("/update-product/{productId}")
+    @PreAuthorize("hasRole('SELLER') and @productService.canModifyProduct(#productId)")
+    public ResponseEntity<ProductEntity> updateProduct(@PathVariable Long productId,
+                                                       @ModelAttribute ProductUpdateRequest request,
+                                                       @RequestParam("thumbnailImage")MultipartFile thumbnailImage,
+                                                       @RequestParam("image") MultipartFile image) throws IOException {
+        ProductEntity updatedProduct = productService.updateProduct(productId, request, thumbnailImage, image);
+
+        return ResponseEntity.ok(updatedProduct);
+    }
+
+    @DeleteMapping("/delete/{productId}")
+    @PreAuthorize("hasRole('SELLER') and @productService.canModifyProduct(#productId)")
+    public ResponseEntity<Void> deleteProduct(@PathVariable Long productId) throws IOException {
+        productService.deleteProduct(productId);
+        return ResponseEntity.noContent().build();
+    }
+
+    @DeleteMapping("/delete-all")
```

Comment:
similar comment as above.
plus, user can only delete their own product, 
so user must be product's seller

Reply:
The user is already related only to his product. SellerEntity currentSeller = sellerService.getCurrentSeller(); -> Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        return (SellerEntity) authentication.getPrincipal();